src/Polly.Core/PredicateBuilder.TResult.cs

```diff
@@ -0,0 +1,140 @@
+    /// <returns>The same instance of the <see cref="PredicateBuilder{TResult}"/> for chaining.</returns>
+    public PredicateBuilder<TResult> Handle<TException>()
+        where TException : Exception
+    {
+        return Handle<TException>(static _ => true);
+    }
+
+    /// <summary>
+    /// Adds a predicate for handling exceptions of the specified type.
+    /// </summary>
+    /// <typeparam name="TException">The type of the exception to handle.</typeparam>
+    /// <param name="predicate">The predicate function to use for handling the exception.</param>
+    /// <returns>The same instance of the <see cref="PredicateBuilder{TResult}"/> for chaining.</returns>
+    /// <exception cref="ArgumentNullException">Thrown when the <paramref name="predicate"/> is <see langword="null"/>.</exception>
+    public PredicateBuilder<TResult> Handle<TException>(Func<TException, bool> predicate)
+        where TException : Exception
+    {
+        Guard.NotNull(predicate);
+
+        return Add(outcome => outcome.Exception is TException exception && predicate(exception));
+    }
+
+    /// <summary>
+    /// Adds a predicate for handling inner exceptions of the specified type.
+    /// </summary>
+    /// <typeparam name="TException">The type of the inner exception to handle.</typeparam>
+    /// <returns>The same instance of the <see cref="PredicateBuilder{TResult}"/> for chaining.</returns>
+    public PredicateBuilder<TResult> HandleInner<TException>()
+        where TException : Exception
+    {
+        return HandleInner<TException>(static _ => true);
+    }
+
+    /// <summary>
+    /// Adds a predicate for handling inner exceptions of the specified type.
+    /// </summary>
+    /// <typeparam name="TException">The type of the inner exception to handle.</typeparam>
+    /// <param name="predicate">The predicate function to use for handling the inner exception.</param>
+    /// <returns>The same instance of the <see cref="PredicateBuilder{TResult}"/> for chaining.</returns>
+    /// <exception cref="ArgumentNullException">Thrown when the <paramref name="predicate"/> is <see langword="null"/>.</exception>
+    public PredicateBuilder<TResult> HandleInner<TException>(Func<TException, bool> predicate)
+        where TException : Exception
+    {
+        Guard.NotNull(predicate);
+
+        return Add(outcome => outcome.Exception?.InnerException is TException innerException && predicate(innerException));
+    }
+
+    /// <summary>
+    /// Adds a predicate for handling results.
+    /// </summary>
+    /// <param name="predicate">The predicate function to use for handling the result.</param>
+    /// <returns>The same instance of the <see cref="PredicateBuilder{TResult}"/> for chaining.</returns>
+    public PredicateBuilder<TResult> HandleResult(Func<TResult, bool> predicate)
+        => Add(outcome => outcome.TryGetResult(out var result) && predicate(result!));
```

Comment:
Can we annotate `TryGetResult` so the compiler knows `result` won't be `null` if it returns `true`? Then we won't need the `!`.

Reply:
The problem is that we should also allow `nulls` there.

You can do `Outcome.FromResult<object?>(null)`. It's perfectly valid result.